File: 112.路劲总和.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int count = 0, s=0;

    void dfs(TreeNode* root, int sum){
        if(!root)
            return;
        s += root->val;
        if(!root->left && !root->right){
            if(s == sum)
                ++count;
            s -= root->val;
            return;
        }
        if(root->left){
            dfs(root->left, sum);
        }
        if(root->right){
            dfs(root->right, sum);
        }
        s -= root->val;
    }

    bool hasPathSum(TreeNode* root, int sum) {
        dfs(root, sum);
        return count;
    }
};

File: 树/112.路劲总和.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
 var hasPathSum = function(root, targetSum) {
    let sum = 0;

    const helper = (root, targetSum) => {
        if (root == null)
            return false;
        sum += root.val;
        if (root.left == null && root.right == null && sum == targetSum) {
            return true;
        }
        if (helper(root.left, targetSum) || helper(root.right, targetSum))
            return true;
        sum -= root.val;
        return false;
    }

    return helper(root, targetSum);
};